### PR TITLE
Ruby on Rails: Associations

### PR DESCRIPTION
**Pull request by mke2111**
For this project, we:-

- Created user, event, and attendee models and controllers.
- Added many to many relationships amongst user, event, and attendee which are attended_events and user_attendee.
- Implemented a lightweight authentication feature.
- Created views to display events and users.

### CODE REVIEW

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 the design is 🤯 you still need to make some changes though! 

## Required Changes ♻️ 

- [x] when I create a user, I get an error in Users#Show, the reason is you're looking for an event using the id params, and I don't have any events _yet_ so it will make that error, remove the `@event` variable to fix this.

- [x] when creating an event, it redirects me to Events#Show and i get an error because you wrote `Attendance.new` instead of `Attendee.new`

- [x] trying to join an event gives me this error page whenever i try to load to the user's page again: 
image

the reason is you're using `@event` (which i deleted to solve the problem, instead you need to use `event` that you have in `@created_events.each do |event|`

- [x] the `home` and `my events` links are not working, and have `#` as the link ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [x] use `foreign_key` and `class_name` in your models
  * [x] User model: `has_many :events, foreign_key: :creator_id` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  * [x] Event model `belongs_to :creator, class_name: "User"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  - [x] events table has creator_id as the foreign key ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

* [x] authentication should have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [x] event managment features should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))


### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def index
+    @users = User.all
+  end
+
+  def show
+    @created_events = current_user.events
+    @user = User.find(params[:id])
+    @event = Event.find(params[:id])
```
Comment (elmejdki) [inline app/controllers/users_controller.rb]: - [ ] how come the user and event have the same ID??, when I try to create a new user an error occur marking this page check the image below:

image

remove the @event variable as mentioned by the previous TSE.

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 

Great job so far @team :+1:, supper cool design though I really liked it :100: :1st_place_medal: . You still have some changes :hammer: :wrench: :toolbox: to take care of in order to get this project approved :checkered_flag:.

## Required Changes:

- [ ] if a user tries to attend the event multiple times, it will duplicate his name on the attendee's list. please hide the I will assist button for users that are already registered, or come up with another solution for it.

image

here we can see user `test2` is duplicated

- [ ] Please add integration tests as mentioned by the previous TSE

- [ ] you have some tests failing please fix them:
image

- [ ] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming:
image

- [ ] add the link that redirects to the users show page the only way that I was able to access this page is using the URL.

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_If you have any questions don't hesitate to contact me on slack._

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Wow, team, an amazing job so far! Really like your design! 👍 
However, there are some requested changes by previous TSE that have not been implemented yet! Let's go with them! 💪 
Once the changes are done, resubmit your project for a new code review ⏩ 
Good luck with it! 😄 
## Mandatory changes
- Add integration test for all the different features, attend and event, unattended an event, login, signup, create event...
- Also unit tests to test models validations
- Also fix the tests that are failing
- The users are still duplicates when accessing multiple times to an event, kindly fix it.

**Comment by edxco** [comment]
Fixed the following:

- A user can't attend twice to the same event
- Add rspec tests

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required :recycle:
#### Hi @mke2111 and @edxco :clap: :clap:
#### Good job so far :+1: I know that you are working hard and doing your best to implement the required changes in order to get this project approved, just keep learning and don't be discouraged about the reviews.
#### Some of the required changes mentioned in the previous review are still missing. Check my comments below :point_down:

#### Tests:
- **_The integration tests are still missing [check this link](https://www.codewithjason.com/rails-integration-tests-rspec-capybara/) and [this one](https://thoughtbot.com/blog/rspec-integration-tests-with-capybara) and [this](https://www.microverse.org/blog/understanding-rspec-controller-request-specs-integration-test-using-capybara) to learn more about how to write integration tests. Add tests for all the different features, attend and event, unattended an event, login, signup, create event_**

- **_Test the association between models(this is not hard you can add `shoulda` gem to make the association test easier)_**

#### _Reviewed by Temitayo Adeleke_

**Review by elmejdki — APPROVED** [review]
# status: approved :heavy_check_mark: 

## Great job @mke2111 :clap: :clap: :clap: :clap:, thanks for making the required changes :100: :+1:.

now you can merge to develop your project is approved.

## Happy Coding :tophat:

**Pull request merged**